python/src/spark_rapids_ml/knn.py

```diff
@@ -590,32 +701,534 @@ async def do_allGather() -> List[str]:
+
+    algorithm = Param(
+        Params._dummy(),
+        "algorithm",
+        "The algorithm to use for approximate nearest neighbors search.",
+        typeConverter=TypeConverters.toString,
+    )
+
+    algoParams = Param(
+        Params._dummy(),
+        "algoParams",
+        "The parameters to use to set up a neighbor algorithm.",
+        typeConverter=DictTypeConverters._toDict,
+    )
+
+    def setAlgorithm(self: P, value: str) -> P:
+        """
+        Sets the value of `algorithm`.
+        """
+        assert value == "ivfflat", "Only IVFFLAT algorithm is currently supported"
+        self._set_params(algorithm=value)
+        return self
+
+    def getAlgorithm(self: P) -> str:
+        """
+        Gets the value of `algorithm`.
+        """
+        return self.getOrDefault("algorithm")
+
+    def setAlgoParams(self: P, value: Dict[str, Any]) -> P:
+        """
+        Sets the value of `algoParams`.
+        """
+        self._set_params(algoParams=value)
+        return self
+
+    def getAlgoParams(self: P) -> Dict[str, Any]:
+        """
+        Gets the value of `algoParams`.
+        """
+        return self.getOrDefault("algoParams")
+
+
+class ApproximateNearestNeighbors(
+    ApproximateNearestNeighborsClass, _CumlEstimator, _ApproximateNearestNeighborsParams
+):
+    """
+    ApproximateNearestNeighbors retrieves k approximate nearest neighbors (ANNs) in item vectors for each query.
+    The key APIs are similar to the NearestNeighbor class which returns the exact k nearest neighbors.
+    The ApproximateNearestNeighbors is currently built on the IVFFLAT algorithm of cuML, and is expected to support
+    other algorithms such as IVFPQ.
+
+    IVFFLAT algorithm trains a set of kmeans centers, then partition every item vector to the closest center. In the query processing
+    phase, a query will be partitioned into a number of closest centers, and probe all the items associated with those centers. In
+    the end the top k closest items will be returned as the approximate nearest neighbors.
+
+    The current implementation build kmeans index independently on every GPU with its part of item vectors. Queries will be broadcast
+    to all GPUs, then every query probes closest centers on individual index. Local topk results will be aggregated to obtain
+    global topk ANNs.
+
+
+    Parameters
+    ----------
+    k: int (default = 5)
+        the default number of approximate nearest neighbors to retrieve for each query.
+
+    algorithm: str (default = 'ivfflat')
+        the algorithm parameter to be passed into cuML. It currently must be 'ivfflat'. Other algorithms such as
+        'ivfpq' are expected to be supported later.
+
+    algoParams: Optional[Dict[str, Any]] (default = None)
+        if set, algoParam is used to configure the algorithm.
+        When algorithm is 'ivfflat':
+            *nlist: (int) number of kmeans clusters to partition the dataframe into.
```

Comment:
Are these two params per GPU or total across all GPUs/partitions?

Reply:
per data partition or batch if max_record_batch used. 
GPUs are assigned in round-robin manner per the current design of CumlModel class.